src/ray/raylet/node_manager.cc

```diff
@@ -1527,41 +1529,45 @@ void NodeManager::FinishAssignedTask(Worker &worker) {
   }
 }
 
-void NodeManager::HandleTaskReconstruction(const TaskID &task_id) {
+void NodeManager::HandleTaskReconstruction(const TaskID &task_id,
+                                           bool return_values_lost) {
   // Retrieve the task spec in order to re-execute the task.
   RAY_CHECK_OK(gcs_client_->raylet_task_table().Lookup(
       JobID::nil(), task_id,
       /*success_callback=*/
-      [this](ray::gcs::AsyncGcsClient *client, const TaskID &task_id,
-             const ray::protocol::TaskT &task_data) {
+      [this, return_values_lost](ray::gcs::AsyncGcsClient *client, const TaskID &task_id,
+                                 const ray::protocol::TaskT &task_data) {
         // The task was in the GCS task table. Use the stored task spec to
         // re-execute the task.
         const Task task(task_data);
-        ResubmitTask(task);
+        ResubmitTask(task, return_values_lost);
       },
       /*failure_callback=*/
-      [this](ray::gcs::AsyncGcsClient *client, const TaskID &task_id) {
+      [this, return_values_lost](ray::gcs::AsyncGcsClient *client,
+                                 const TaskID &task_id) {
         // The task was not in the GCS task table. It must therefore be in the
         // lineage cache.
         RAY_CHECK(lineage_cache_.ContainsTask(task_id));
         // Use a copy of the cached task spec to re-execute the task.
         const Task task = lineage_cache_.GetTask(task_id);
-        ResubmitTask(task);
-
+        ResubmitTask(task, return_values_lost);
       }));
 }
 
-void NodeManager::ResubmitTask(const Task &task) {
+void NodeManager::ResubmitTask(const Task &task, bool return_values_lost) {
   if (task.GetTaskSpecification().IsActorTask()) {
     // Actor reconstruction is turned off by default right now.
     const ActorID actor_id = task.GetTaskSpecification().ActorId();
     auto it = actor_registry_.find(actor_id);
     RAY_CHECK(it != actor_registry_.end());
     if (it->second.IsAlive()) {
+      // Only treat the task as failed if its output has been evicted.
+      // Otherwise, this must be a spurious reconstruction.
+      if (return_values_lost) {
```

Comment:
What would happen if we unconditionally treated the task as failed? It seems like it would simplify the code a lot to not have to track whether the return values were lost.

Reply:
The code used to do what you're saying, but it was causing spurious `RayGetError`s that we fixed in #3359.

One option that I was thinking we could do is to check whether the task has already been executed using the task counters for the actor, then treat the task as failed if that check passes. To be safe, we could also check whether the return values exist on any nodes, but we wouldn't need the extra logic in this PR to check whether the return values were evicted. What do you think?

Reply:
That sounds reasonable, so it would just involve checking the locations table after the task count check?

Reply:
Yup.

Reply:
Hmm actually, it may still make sense to have the check that the object is evicted. Since GCS writes are asynchronous, the locations entry might be empty even when some node does have the object.